src/frequenz/dispatch/_dispatch.py

```diff
@@ -0,0 +1,226 @@
+# License: MIT
+# Copyright © 2024 Frequenz Energy-as-a-Service GmbH
+
+"""Dispatch type with support for next_run calculation."""
+
+
+import logging
+from dataclasses import dataclass
+from datetime import datetime, timezone
+from typing import Iterator, cast
+
+from dateutil import rrule
+from frequenz.client.dispatch.types import Dispatch as BaseDispatch
+from frequenz.client.dispatch.types import Frequency, Weekday
+
+_logger = logging.getLogger(__name__)
+"""The logger for this module."""
+
+_RRULE_FREQ_MAP = {
+    Frequency.MINUTELY: rrule.MINUTELY,
+    Frequency.HOURLY: rrule.HOURLY,
+    Frequency.DAILY: rrule.DAILY,
+    Frequency.WEEKLY: rrule.WEEKLY,
+    Frequency.MONTHLY: rrule.MONTHLY,
+}
+"""To map from our Frequency enum to the dateutil library enum."""
+
+_RRULE_WEEKDAY_MAP = {
+    Weekday.MONDAY: rrule.MO,
+    Weekday.TUESDAY: rrule.TU,
+    Weekday.WEDNESDAY: rrule.WE,
+    Weekday.THURSDAY: rrule.TH,
+    Weekday.FRIDAY: rrule.FR,
+    Weekday.SATURDAY: rrule.SA,
+    Weekday.SUNDAY: rrule.SU,
+}
+"""To map from our Weekday enum to the dateutil library enum."""
+
+
+@dataclass(frozen=True)
+class Dispatch(BaseDispatch):
+    """Dispatch type with extra functionality."""
+
+    deleted: bool = False
+    """Whether the dispatch is deleted."""
+
+    running_state_change_synced: datetime | None = None
+    """The last time a message was sent about the running state change."""
+
+    def __init__(
+        self,
+        client_dispatch: BaseDispatch,
+        deleted: bool = False,
+        running_state_change_synced: datetime | None = None,
+    ):
+        """Initialize the dispatch.
+
+        Args:
+            client_dispatch: The client dispatch.
+            deleted: Whether the dispatch is deleted.
+            running_state_change_synced: Timestamp of the last running state change message.
+        """
+        super().__init__(**client_dispatch.__dict__)
```

Comment:
I mean, it's the same, dataclasses specifically set the dict attribute

Reply:
It is less hacky and I wouldn't be surprised if there is more magic going on. I don't see any reason to use the hacky approach if the official way is enough :)

Reply:
Waiting for feedback on this one, I really think we should use the publicly exposed functions, in python one should usually not use symbols starting with `__` unless it is really necessary. `__dict__` is not even documented in https://docs.python.org/3/library/dataclasses.html.

Reply:
I tried using it and it actually breaks the tests because it converts sub-members also to dicts...

```
Pdb) client_dispatch.__dict__
{'id': 1, 'microgrid_id': 26, 'type': '99346', 'start_time': datetime.datetime(2024, 5, 17, 10, 45, 58, tzinfo=datetime.timezone.utc), 'duration': datetime.timedelta(days=5, seconds=9001), 'selector': [66, 63, 52, 39, 62], 'active': True, 'dry_run': True, 'payload': {'key_3': 'c', 'key_0': 75.58042041572239, 'key_1': True, 'key_2': False}, 'recurrence': RecurrenceRule(frequency=<Frequency.WEEKLY: 4>, interval=67, end_criteria=None, byminutes=[45, 52, 50, 42, 40, 0], byhours=[15, 10, 7, 23, 10, 22, 2, 6, 18], byweekdays=[<Weekday.TUESDAY: 2>, <Weekday.SUNDAY: 7>, <Weekday.TUESDAY: 2>], bymonthdays=[15, 3, 3, 11, 29, 17, 30, 16], bymonths=[5]), 'create_time': datetime.datetime(2024, 5, 6, 15, 27, 50, 359784, tzinfo=datetime.timezone.utc), 'update_time': datetime.datetime(2024, 5, 6, 15, 27, 50, 359784, tzinfo=datetime.timezone.utc)}
(Pdb) dataclasses.asdict(client_dispatch)
{'id': 1, 'microgrid_id': 26, 'type': '99346', 'start_time': datetime.datetime(2024, 5, 17, 10, 45, 58, tzinfo=datetime.timezone.utc), 'duration': datetime.timedelta(days=5, seconds=9001), 'selector': [66, 63, 52, 39, 62], 'active': True, 'dry_run': True, 'payload': {'key_3': 'c', 'key_0': 75.58042041572239, 'key_1': True, 'key_2': False}, 'recurrence': {'frequency': <Frequency.WEEKLY: 4>, 'interval': 67, 'end_criteria': None, 'byminutes': [45, 52, 50, 42, 40, 0], 'byhours': [15, 10, 7, 23, 10, 22, 2, 6, 18], 'byweekdays': [<Weekday.TUESDAY: 2>, <Weekday.SUNDAY: 7>, <Weekday.TUESDAY: 2>], 'bymonthdays': [15, 3, 3, 11, 29, 17, 30, 16], 'bymonths': [5]}, 'create_time': datetime.datetime(2024, 5, 6, 15, 27, 50, 359784, tzinfo=datetime.timezone.utc), 'update
_time': datetime.datetime(2024, 5, 6, 15, 27, 50, 359784, tzinfo=datetime.timezone.utc)}
```

clearly `asdict()` is doing too much magic :P

Reply:
Mmm, OK, I'm still not happy about using an undocumented method starting with `__` but if replacing it is not trivial I'm OK with leaving it *temporarily* (i.e. until all blows up because `__dict__` is removed :laughing: ).